keras_hub/src/models/moonshine/moonshine_decoder.py

```diff
@@ -0,0 +1,486 @@
+        context = keras.layers.Input(shape=[None, hidden_dim])
+        rot_pos_emb = keras.layers.Input(shape=[None, None], batch_size=1)
+        rot_pos_emb = keras.ops.squeeze(rot_pos_emb)
+
+        outputs = self([inputs, context, rot_pos_emb], use_cache=False)
+
+        return keras.Model(
+            inputs=[inputs, context, rot_pos_emb],
+            outputs=outputs,
+        )
+
+    def get_cached_call(self, hidden_dim, key_dim, num_heads):
+        inputs = keras.layers.Input(shape=[None, hidden_dim])
+        context = keras.layers.Input(shape=[None, hidden_dim])
+        cache_k = keras.layers.Input(shape=[None, num_heads, key_dim])
+        cache_v = keras.layers.Input(shape=[None, num_heads, key_dim])
+        x_attn_cache_k = keras.layers.Input(shape=[None, num_heads, key_dim])
+        x_attn_cache_v = keras.layers.Input(shape=[None, num_heads, key_dim])
+        rot_pos_emb = keras.layers.Input(shape=[None, None], batch_size=1)
+        rot_pos_emb = keras.ops.squeeze(rot_pos_emb)
+
+        outputs = self(
+            [
+                inputs,
+                context,
+                cache_k,
+                cache_v,
+                x_attn_cache_k,
+                x_attn_cache_v,
+                rot_pos_emb,
+            ],
+            use_cache=True,
+        )
+
+        return keras.Model(
+            inputs=[
+                inputs,
+                context,
+                cache_k,
+                cache_v,
+                x_attn_cache_k,
+                x_attn_cache_v,
+                rot_pos_emb,
+            ],
+            outputs=outputs,
+        )
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "hidden_dim": self.hidden_dim,
+                "inner_dim": self.inner_dim,
+                "num_heads": self.num_heads,
+                "ff_mult": self.ff_mult,
+                "ff_swiglu": self.ff_swiglu,
+                "pad_head_dim_to_multiple_of": self.pad_head_dim_to_multiple_of,
+            }
+        )
+        return config
+
+
+@keras.saving.register_keras_serializable(package="keras_hub")
+class MoonshineDecoder(keras.Model):
+    """
+    Moonshine decoder.
+
+    A transformer decoder model that stacks multiple MoonshineDecoderBlock
+    layers, an embedding layer with reversible projection, rotary positional
+    embeddings, and a final normalization to produce output logits. This model
+    supports both cached and uncached operation modes for efficient
+    autoregressive generation.
+
+    Args:
+        num_layers (int): The number of decoder layers.
+        hidden_dim (int): The dimensionality of the model.
+        inner_dim (int): The inner dimensionality for the feedforward layers in
+        each decoder block.
+        num_heads (int): The number of attention heads.
+        vocab_size (int): The size of the vocabulary for reversible embeddings.
```

Comment:
rename to `vocabulary_size` to be consistent with other implementations in KerasHub